### PR TITLE
rocAL - performance PR

### PR DESCRIPTION
**Pull request by SundarRajan28**
- Added single handle creation for all augmentations in amd_openvx_extensions
- Added a graph attribute called num_threads which is set in master_graph during graph creation
- The num_threads value is set with the value passed from user during the pipeline creation
- Added a num_threads calculation which calculates the optimal threads to be used for a system
- Uses the optimal num_threads value for OpenMP parallelization during decoding
- The optimal num_threads value is also set in the graph and used for RPP handle creation if the user provided threads is lesser than 0
- Removed internal_batch_size calculation from master graph and all its associated variables and APIs
- Made changes to output routine and output routine video to use user_batch_size for processing
- Added an option in copy_out_tensor for Host backend to use HIP normalization and copy kernel
- Added an option in rocalGetImageLabels for copying to device memory directly even for Host backend
- Added two arguments to pipeline for the above two options in rocAL
- The two arguments - normalization_on_device and copy_labels_to_device is set to False by default so the change won't break existing rocAL loaders
- This rocAL PR is tested with the corresponding RPP PR [#143](https://github.com/GPUOpen-ProfessionalCompute-Libraries/rpp/pull/143)

### CODE REVIEW

**Comment by LakshmiKumar23** [comment]
@SundarRajan28 can you check the codacy errors?

File: amd_openvx_extensions/amd_rpp/include/internal_rpp.h [hunk 1]
```diff
@@ -60,14 +60,14 @@ struct RPPCommonHandle {
 #elif ENABLE_HIP
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/include/internal_rpp.h:60]: Can you please change this structure name to vxRppHandle instead? RPPCommonHandle and RppHandle is confusing

File: amd_openvx_extensions/amd_rpp/source/ExposurebatchPD.cpp [hunk 2]
```diff
@@ -132,22 +131,22 @@ static vx_status VX_CALLBACK processExposurebatchPD(vx_node node, const vx_refer
         refreshExposurebatchPD(node, parameters, num, data);
         if (df_image == VX_DF_IMAGE_U8)
         {
-            rpp_status = rppi_exposure_u8_pln1_batchPD_gpu((void *)data->cl_pSrc, data->srcDimensions, data->maxSrcDimensions, (void *)data->cl_pDst, data->exposureValue, data->nbatchSize, data->rppHandle);
+            rpp_status = rppi_exposure_u8_pln1_batchPD_gpu((void *)data->cl_pSrc, data->srcDimensions, data->maxSrcDimensions, (void *)data->cl_pDst, data->exposureValue, data->nbatchSize, data->handle->rppHandle);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/ExposurebatchPD.cpp:134]: I think it is complaining because data->handle->rppHandle is ambiguous.  (data->handle)->rppHandle and proper typecasting is needed

File: rocAL/rocAL/source/api/rocal_api_meta_data.cpp [hunk 3]
```diff
@@ -231,8 +231,16 @@ ROCAL_API_CALL rocalGetImageLabels(RocalContext p_context, void* buf, unsigned i
     if(context->user_batch_size() != meta_data_batch_size)
         THROW("meta data batch size is wrong " + TOSTR(meta_data_batch_size) + " != "+ TOSTR(context->user_batch_size() ))
 
-    if (context->affinity == RocalAffinity::CPU)
-        memcpy(buf, meta_data.second->get_label_batch().data(), sizeof(int) * meta_data_batch_size);
+    if (context->affinity == RocalAffinity::CPU) {
```
Comment (rrawther) [inline rocAL/rocAL/source/api/rocal_api_meta_data.cpp:234]: Not sure why this is changed from tot version. We need to keep all 3 cases (CPU only, GPU_HIP and GPU_OpenCL). Please revert back to tot

File: rocAL/rocAL/source/loaders/image/image_read_and_decode.cpp [hunk 4]
```diff
@@ -99,6 +99,12 @@ ImageReadAndDecode::create(ReaderConfig reader_config, DecoderConfig decoder_con
         }
     }
     _reader = create_reader(reader_config);
+
```
Comment (rrawther) [inline rocAL/rocAL/source/loaders/image/image_read_and_decode.cpp:102]: There is no need to calculate this here. This has be decided by the mastergraph and send it to decoder via DecoderConfig

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 5]
```diff
@@ -255,11 +254,23 @@ MasterGraph::decrease_image_count()
         _remaining_count -= (_is_sequence_reader_output ? _sequence_batch_size : _user_batch_size);
 }
 
+void
+MasterGraph::calculate_cpu_num_threads(size_t shard_count)
+{
+    if (_cpu_num_threads <= 0) {
+        const unsigned DEFAULT_SMT_COUNT = 2;
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:261]: please use standard naming convension throughout (rename all CAP names)

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 6]
```diff
@@ -128,9 +130,9 @@ def __init__(self, batch_size=-1, num_threads=-1, device_id=-1, seed=1,
         self._default_cuda_stream_priority = default_cuda_stream_priority
         self._tensor_layout = tensor_layout
         self._tensor_dtype = tensor_dtype
-        self._multiplier = multiplier
+        self._multiplier = multiplier if multiplier else [1.0,1.0,1.0]
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py:133]: I think original code should work for handling default value

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 7]
```diff
@@ -181,7 +183,7 @@ def copyImage(self, array):
     def copyToTensor(self, array,  multiplier, offset, reverse_channels, tensor_format, tensor_dtype):
 
         b.rocalCopyToOutputTensor(self._handle, ctypes.c_void_p(array.data_ptr()), tensor_format, tensor_dtype,
-                                    multiplier[0], multiplier[1], multiplier[2], offset[0], offset[1], offset[2], (1 if reverse_channels else 0))
+                                    multiplier[0], multiplier[1], multiplier[2], offset[0], offset[1], offset[2], (1 if reverse_channels else 0), (1 if self._normalization_on_device else 0))
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py:186]: I think we need to discuss how to support this for CPU pipeline. 
we need to allocate pinned memory if the output requested is on device. That way, we can eliminate extra host to device copy.

**Review by rrawther — CHANGES_REQUESTED** [review]
@SundarRajan28 : I would like to have a call on this to discuss the optimal way to copy output.

File: rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py [hunk 8]
```diff
@@ -88,39 +90,40 @@ def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = Fals
         color_format = b.getOutputColorFormat(self.loader._handle)
         self.p = (1 if (color_format == int(types.GRAY)) else 3)
         self.labels_size = ((self.bs*self.loader._numOfClasses) if (self.loader._oneHotEncoding == True) else self.bs)
-        if tensor_layout == types.NCHW:
-            if self.device == "cpu":
+        if self.tensor_format == types.NCHW:
+            if self.device == "cpu" and not self.normalization_on_device:
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:94]: this will break the flow when there is multiple outputs from the pipeline.

File: amd_openvx_extensions/amd_rpp/include/internal_rpp.h [hunk 9]
```diff
@@ -60,14 +60,14 @@ struct RPPCommonHandle {
...
-    void* cpuHandle = NULL;
+    rppHandle_t rppHandle;
     int count;
 };
 
 //! Brief The utility functions
 vx_node createNode(vx_graph graph, vx_enum kernelEnum, vx_reference params[], vx_uint32 num);
-vx_status createGraphHandle(vx_node node, RPPCommonHandle ** pHandle);
-vx_status releaseGraphHandle(vx_node node, RPPCommonHandle * handle);
+vx_status createGraphHandle(vx_node node, RPPCommonHandle ** pHandle, Rpp32u batchSize, Rpp32u deviceType);
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply amd_openvx_extensions/amd_rpp/include/internal_rpp.h]: Resolved in the latest commit

File: amd_openvx_extensions/amd_rpp/include/internal_rpp.h (hunk 1 above)
Comment (SundarRajan28) [reply amd_openvx_extensions/amd_rpp/include/internal_rpp.h:60]: Renamed the structure in the latest commit

File: rocAL/rocAL/source/api/rocal_api_meta_data.cpp (hunk 3 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/api/rocal_api_meta_data.cpp:234]: This change was added since we added a memcpy_to_device parameter in pipeline which if set, passes a positive value to the flags variable in this function. We are checking for that value and if its set, assumes the user is asking to copy to a HIP buffer hence added an else block with hipMemcpy. I will add enable_hip and enable_opencl flags inside the else block to support GPU_HIP and GPU_OCL

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 6 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py:133]: There was a codacy warning which warned about having mutable default values in init function. I changed the init function to fix the warning

File: rocAL/rocAL/source/api/rocal_api_meta_data.cpp (hunk 3 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/api/rocal_api_meta_data.cpp:234]: Added check for HIP/OCL backend and copying labels to buffers accordingly

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 5 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in the latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 10]
```diff
@@ -255,11 +254,23 @@ MasterGraph::decrease_image_count()
...
+        unsigned THREAD_COUNT = std::thread::hardware_concurrency();
+        size_t CORE_COUNT = THREAD_COUNT / DEFAULT_SMT_COUNT;
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in the latest commit

File: rocAL/rocAL/source/loaders/image/image_read_and_decode.cpp (hunk 4 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/loaders/image/image_read_and_decode.cpp]: Resolved in the latest commit

File: rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py (hunk 8 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py]: Changed the tensor creation according to the output memory type as discussed in the call

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 7 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py]: Changed the API calls according to the changes discussed in the call

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 6 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py]: Passing the mean and std to pipeline and calculating the multiplier and offset in pipeline as discussed in the call

File: rocAL/rocAL/include/api/rocal_api_types.h [hunk 11]
```diff
@@ -137,6 +137,13 @@ enum RocalDecoderType
     ROCAL_DECODER_VIDEO_FFMPEG_HW = 4
 };
 
+enum RocalOutputMemType
+{
+    ROCAL_MEMCPY_HOST = 0,
```
Comment (rrawther) [inline rocAL/rocAL/include/api/rocal_api_types.h:142]: Is the defines in 148 to 150 replicating the same functionality? Please remove the one which is not used

Comment (SundarRajan28) [reply rocAL/rocAL/include/api/rocal_api_types.h:142]: Those defines are used in getImageLabels API and similar metadata API functions to force return the metadata in GPU memory if set. I tried using the already present defines but I wasn't able to add python bindings for defines so I added an enum.

**Comment by kiritigowda** [comment]
@SundarRajan28 can you look at these 2 issues on codacy - https://github.com/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pull/1060/checks?check_run_id=12873006986

**Comment by SundarRajan28** [comment]
@rrawther The remaining codacy warning in SequenceRearrange.cpp is due to the handle being only used inside ENABLE_OPENCL block and not used anywhere else. Due to this, codacy is throwing warning that the handle variable is not used anywhere.

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp [hunk 12]
```diff
@@ -1925,36 +1925,50 @@ vx_node createNode(vx_graph graph, vx_enum kernelEnum, vx_reference params[], vx
     return node;
 }
 
-// looks like following code is not used
-#if 0//ENABLE_OPENCL
-
-vx_status createGraphHandle(vx_node node, RPPCommonHandle **pHandle)
-{
-    RPPCommonHandle *handle = NULL;
+vx_status createRPPHandle(vx_node node, vxRppHandle **pHandle, Rpp32u batchSize, Rpp32u deviceType) {
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:1928]: adjust the spacing here

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp [hunk 13]
```diff
@@ -1925,36 +1925,50 @@ vx_node createNode(vx_graph graph, vx_enum kernelEnum, vx_reference params[], vx
...
+    vxRppHandle *handle = NULL;
     STATUS_ERROR_CHECK(vxGetModuleHandle(node, OPENVX_KHR_RPP, (void **)&handle));
-    if (handle)
-    {
+    vx_uint32 cpu_num_threads;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:1931]: please follow spacing requirement according to google coding standard

File: rocAL/rocAL/include/api/rocal_api_meta_data.h [hunk 14]
```diff
@@ -123,7 +123,7 @@ extern "C" unsigned ROCAL_API_CALL rocalGetImageNameLen(RocalContext rocal_conte
 
 /// \param meta_data RocalMetaData object that contains info about the images and labels
 /// \param buf user's buffer that will be filled with labels. Its needs to be at least of size batch_size.
-extern "C" void ROCAL_API_CALL rocalGetImageLabels(RocalContext rocal_context, void* buf,  unsigned int flags=0);
+extern "C" void ROCAL_API_CALL rocalGetImageLabels(RocalContext rocal_context, void* buf,  RocalOutputMemType output_mem_type);
```
Comment (rrawther) [inline rocAL/rocAL/include/api/rocal_api_meta_data.h:126]: not sure why the flags is removed

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 15]
```diff
@@ -178,43 +179,43 @@ def copyImage(self, array):
         b.rocalCopyToOutput(
             self._handle, np.ascontiguousarray(out, dtype=array.dtype))
 
-    def copyToTensor(self, array,  multiplier, offset, reverse_channels, tensor_format, tensor_dtype):
+    def toTensor(self, array,  multiplier, offset, reverse_channels, tensor_format, tensor_dtype):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py:182]: We still need to define this fn as copyToExternalTensor() or something similar since the api does the tensor conversion as well copying to external memory. Applies to all similar functions

File: rocAL/rocAL_pybind/examples/rocAL_api_coco_pipeline.py [hunk 16]
```diff
@@ -287,29 +291,19 @@ def coco_anchors():  # Should be Tensor of floats in ltrb format - input - Mx4 w
             res_images, saturation=saturation, contrast=contrast, brightness=brightness, hue=hue)
         flip_coin = fn.random.coin_flip(probability=0.5)
         bboxes = fn.bb_flip(bboxes, ltrb=True, horizontal=flip_coin)
-        if args.display:
-            cmn_images = fn.crop_mirror_normalize(ct_images,
+        cmn_images = fn.crop_mirror_normalize(ct_images,
```
Comment (rrawther) [inline rocAL/rocAL_pybind/examples/rocAL_api_coco_pipeline.py:294]: is the cmn output float here?

**Review by rrawther — CHANGES_REQUESTED** [review]
@SundarRajan28 please merge all upstream changes

File: rocAL/rocAL/include/api/rocal_api_meta_data.h (hunk 14 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/api/rocal_api_meta_data.h:126]: The flags are replaced by the new Enum - the flags were used to force copy to GPU memory from CPU if set. We are doing the same when setting GPU_MEMORY as the RocalOutputMemType

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp (hunk 12 above)
Comment (SundarRajan28) [reply amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:1928]: I am not seeing any spacing issues in this line

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 15 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py]: Renamed all the python API functions to copyToExternalTensor(). Still keeping the rocAL API functions as toTensor

File: rocAL/rocAL_pybind/examples/rocAL_api_coco_pipeline.py (hunk 16 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/rocAL_api_coco_pipeline.py:294]: Modified all CMN nodes in python examples to use NHWC/U8 format by default

**Comment by SundarRajan28** [comment]
@rrawther Resolved the new review comments and merged the PR branch with upstream develop changes

File: rocAL/rocAL_pybind/setup.py [hunk 17]
```diff
@@ -18,72 +18,29 @@
...
-      description='AMD ROCAL',
-      url='https://github.com/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/tree/master/rocal',
-      version='1.1.0',
+setup(
+      name='amd-rocal',
+      description='AMD ROCm Augmentation Library',
+      url='https://github.com/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/rocAL',
+      version='1.0.0',
       author='AMD',
       license='Apache License 2.0',
```
Comment (rrawther) [inline rocAL/rocAL_pybind/setup.py:40]: do we want to say apache Licence since others are MIT

File: rocAL/rocAL_pybind/setup.py [hunk 18]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2018 - 2023 Advanced Micro Devices, Inc. All rights reserved.
+# Copyright (c) 2018 - 2022 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (rrawther) [inline rocAL/rocAL_pybind/setup.py:1]: why is this file included in this PR?

File: rocAL/rocAL_pybind/getrocALwheelname.py [hunk 19]
```diff
@@ -0,0 +1,20 @@
+from setuptools.dist import Distribution
```
Comment (rrawther) [inline rocAL/rocAL_pybind/getrocALwheelname.py:1]: Looks like there are already merged with tot?

**Comment by LakshmiKumar23** [comment]
@SundarRajan28 This PR seems to be created to the master. Plus it's not synced with TOT. Can you please sync?

File: rocAL/rocAL_pybind/examples/tf_petsTrainingExample/train_withROCAL_withTFRecordReader.py [hunk 20]
```diff
@@ -142,13 +142,14 @@ def main():
                                               mean=[0,0,0],
```
Comment (rrawther) [inline rocAL/rocAL_pybind/examples/tf_petsTrainingExample/train_withROCAL_withTFRecordReader.py:142]: Are these the right values for mean and std?

**Review by rrawther — APPROVED** [review]
@LakshmiKumar23 to verify the PR
@kiritigowda: some tot changes are showing as file changes in this PR. Don't know what it is comparing with

File: rocAL/rocAL_pybind/examples/tf_petsTrainingExample/train_withROCAL_withTFRecordReader.py (hunk 20 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/examples/tf_petsTrainingExample/train_withROCAL_withTFRecordReader.py:142]: The mean and std for the CMN was 0 and 255 so I added the same values for pipeline initialization

**Comment by SundarRajan28** [comment]
@LakshmiKumar23 The PR is created to the develop branch and has been merged with the latest develop changes

File: rocAL/rocAL_pybind/getrocALwheelname.py (hunk 19 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/getrocALwheelname.py]: Since PR destination branch was changed to master which didn't want the pybind changes, this was showing up as files changed. The PR destination branch is changed to develop and merged with latest changes.

File: rocAL/rocAL_pybind/setup.py (hunk 18 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/setup.py]: Since PR destination branch was changed to master which didn't want the pybind changes, this was showing up as files changed. The PR destination branch is changed to develop and merged with latest changes.

File: rocAL/rocAL_pybind/setup.py (hunk 17 above)
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/setup.py]: Since PR destination branch was changed to master which didn't want the pybind changes, this was showing up as files changed. The PR destination branch is changed to develop and merged with latest changes.

**Review by LakshmiKumar23 — APPROVED** [review]

**Comment by LakshmiKumar23** [comment]
@kiritigowda Can be merged along with PR143 on RPP

**Comment by rrawther** [comment]
@kiritigowda : Can you please merge this PR. I guess you mentioned about the setup script change. Please let @SundarRajan28  know what he need to do

**Comment by kiritigowda** [comment]
> @kiritigowda : Can you please merge this PR. I guess you mentioned about the setup script change. Please let @SundarRajan28 know what he need to do

@rrawther I have merged the RPP changes into this branch, waiting for CI to pass to merge.
